### DISCUSSION TITLE
understanding the "translate()" functionality

### DISCUSSION THREAD

**Discussion by jbernalr** [CadQuery/cadquery#1185] [Q&A] [answered]
I have here a minimal code: 

```python
import cadquery as cq
from cadquery import exporters

filename = './rack.stl'
l_rack = 30 # mm [lenght of the rack]
base_x = 10  # mm [base width]
base_y = 10  # mm [base height]

r_x = 0.6
r_y = 0.6 

rack = cq.Workplane("XY").box(base_x, base_y, l_rack).\
    faces(">Z").workplane().rect(r_x*base_x, r_y*base_y).cutThruAll().\
    workplane().translate((0,50,10)).rect(r_x*base_x, r_y*base_x).extrude(10)

```

which produces the figure depicted at the bottom. 
however, I cannot translate my extrusion from the center. 

I tried the following variations without success: 

`workplane().translate((0,50,10)).rect(r_x*base_x, r_y*base_x).extrude(10)`
`faces(">Z").workplane().translate((0,50,10)).rect(r_x*base_x, r_y*base_x).extrude(10)`
`workplane().rect(r_x*base_x, r_y*base_x).translate((0,50,10)).extrude(10)`
`workplane().rect(r_x*base_x, r_y*base_x).extrude(10).translate((0,50,10)).`

I read the docs and saw examples where the translation happens at the trailing part of the object. 
Is this behaviour as expected? can anyone spot the error? 

<img width="898" alt="Screenshot 2022-11-02 at 15 42 22" src="https://user-images.githubusercontent.com/33322694/199519568-c8f18523-356a-43b0-987c-84e6b4f9dd7f.png">

**Comment by lorenzncode** [comment]
`workplane()` was called immediately before `translate((0, 50, 10))`.   This is a problem.   

translate _"Returns a copy of all of the items on the stack moved by ..."_

By calling `workplane()` you create an empty workplane so there are no items on the stack to move.

Here is an example where the box is moved:
```py
rack = (
    cq.Workplane("XY")
    .box(base_x, base_y, l_rack)
    .translate((0, 50, 10))
)
```

**Reply by jbernalr** [reply]
@lorenzncode thanks. Just let say you are r building (as in my case) on top of this initial box of you and pick one of the faces to create a square you want to extrude. Where should the translate be located? I can reproduce your example, but I cannot use translate on a face of an existing body.
<img width="579" alt="Screenshot 2022-11-03 at 15 49 25" src="https://user-images.githubusercontent.com/33322694/199753827-b3818200-598f-4c2a-a292-99f0b61f84bc.png">

**Comment by lorenzncode** [comment, marked as answer]
The origin depends on centerOption when creating a workplane on a face.  One way to change the origin is to specify it (requires "ProjectedOrigin" centerOption which is the default).

```py
result = (
    cq.Workplane()
    .box(10, 10, 30)
    .faces(">Z")
    .workplane(origin=(0, 2, 0))
    .rect(6, 6)
    .extrude(-10, combine="s")
)
```


Or call `center`:

```py
result = (
    cq.Workplane()
    .box(10, 10, 30)
    .faces(">Z")
    .workplane()
    .center(0, 2)
    .rect(6, 6)
    .extrude(-10, combine="s")
)
```

Another way to locate a workplane is covered in the docs example "Locating a Workplane on a vertex".  Here locating on an edge:

```py
result = (
    cq.Workplane()
    .box(10, 10, 30)
    .faces(">Z")
    .edges(">Y")
    .workplane(centerOption="CenterOfMass")
    .center(0, -3)
    .rect(6, 6)
    .extrude(-10, combine="s")
)
```

You might have moved the shape in a sketch:

```py
s = cq.Sketch().push([(0, 2)]).rect(6, 6)

result = (
    cq.Workplane()
    .box(10, 10, 30)
    .faces(">Z")
    .workplane()
    .placeSketch(s)
    .extrude(-10, combine="s")
)
```


(Note that `Sketch` and `extrude` with combine="s" is not available in version 2.1.)

**Comment by jbernalr** [comment]
@lorenzncode that's it. thanks!
would be correct to say that translate only works on solids created from the cadquery primitives?
I may do a PR with a few sentences in the documentation

**Comment by lorenzncode** [comment]
We can create the same shape as before using translate (probably not the most straightforward way):

```py
result = (
    cq.Workplane()
    .box(10, 10, 30)
    .faces(">Z")
    .translate((0, 2, 0))
    .wires()
    .toPending()
    .offset2D(-2)
    .extrude(-10, combine="s")
)
```

translate works with any type of shape - a face in this example.
